martini_sour/src/analyze.py

```diff
@@ -0,0 +1,175 @@
+# Copyright 2021 University of Groningen
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+import MDAnalysis as mda
+from tqdm import tqdm
+import matplotlib.pyplot as plt
+from scipy.optimize import curve_fit
+from .titrate import get_ph_values
+
+
+def setup_dirs(ph_range, sim_dir, traj_file, tpr_file):
+    """
+    Sets up directory paths based on whether multidir options have been provided or not.
+    """
+
+    if ph_range and sim_dir:
+        ph_values = get_ph_values(ph_range)
+        traj_paths = [f'{ph}/{sim_dir}/{traj_file}' for ph in ph_values]
+        tpr_paths = [f'{ph}/{sim_dir}/{tpr_file}' for ph in ph_values]
+    else:
+        traj_paths = [traj_file]
+        tpr_paths = [tpr_file]
+
+    return traj_paths, tpr_paths, np.array(ph_values, dtype=float)
+
+
+def save_to_file(out_file, ph_values, props, sel_commands):
+    array = np.concatenate((ph_values[np.newaxis], props.T), axis=0)
+
+    fmt = ['%6.2f']+['%10.5f']*(len(array)-1)
+    header = '  pH ' + ' '.join([f'sel. {i}'.rjust(10) for i in range(1, len(array))])
+
+    np.savetxt(out_file, array.T, fmt=fmt, header=header)
+
+
+def do_statistics(degrees_of_deprot):
+    """
+    This is probably where a more complicated statistical analysis would be inferfaced
+    """
+
+    mean = np.array(degrees_of_deprot).mean(axis=-1)
+    std = np.array(degrees_of_deprot).std(axis=-1)
+
+    return mean, std
+
+
+def calc_degree_of_deprot(ph, pka, q):
+    return 1/(10**(q*(pka-ph))+1)
+
+
+def do_plotting(out_file, mean, ph_values):
+    ph_range = np.arange(ph_values.min(), ph_values.max()+1e-3, 0.1)
+    fig = plt.figure()
+
+    for i, one_sel in enumerate(mean.T, start=1):
+
+        fit = curve_fit(calc_degree_of_deprot, ph_values, one_sel, p0=(7.5, 0.75))[0]
```

Comment:
When you do curve-fit you actually loose all the information about the standard error in the titration, which actually can be quite important when they are significant. Furthermore, there is no error computed with the fit values. Finally, it has proven more robust to be able to guess initial values for the pKa for when pKa curves deviate from their ideal behavior.  

The [symfit package](https://symfit.readthedocs.io/en/stable/) actually allows you to do fitting with errors taken into account and it computes the error in the fitted values as well. I'd recommend to use that function instead of curve fit. Otherwise we should at least compute the errors in the fitted parameters from the covariance matrix that should be given out by curve-fite.